File: backend-springboot/todo/src/main/java/net/indralight/todo/service/UserService.java
package net.indralight.todo.service;

import org.springframework.beans.factory.annotation.Autowired;
//import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import lombok.extern.slf4j.Slf4j;
import net.indralight.todo.dto.KakaoOAuthDTO;
import net.indralight.todo.dto.SnsUserDTO;
import net.indralight.todo.model.UserEntity;
import net.indralight.todo.persistence.UserRepository;

@Slf4j
@Service
public class UserService {
    @Autowired
    private UserRepository userRepository;

    @Autowired
    private KakaoOAuthService oauthService;

    public UserEntity createByLocal(final UserEntity userEntity) {

        if (userEntity == null || userEntity.getEmail() == null) {
            throw new RuntimeException("Invalid arguments");
        }
        final String email = userEntity.getEmail();
        final String provider = userEntity.getProvider();
        if (userRepository.existsByEmailAndProvider(email, provider)) {
            log.warn("Email already exists {}", email);
            throw new RuntimeException("Email already exists");
        }
        return userRepository.save(userEntity);
    }

//    public UserEntity getByCredentials(final String email, final String password, final PasswordEncoder encoder) {
//
//        final UserEntity originalUser = userRepository.findByEmail(email);
//        if (originalUser != null && encoder.matches(password, originalUser.getPassword())) {
//            return originalUser;
//        }
//
//        return null;
//    }

    public UserEntity getByLocal(final String email, final String password) {
        final String provider = "local";

        final UserEntity originalUser = userRepository.findByEmailAndProvider(email, provider);
        if (originalUser != null && password.equals(originalUser.getPassword())) {
            return originalUser;
        }

        return null;
    }

    public UserEntity getByKakao(KakaoOAuthDTO kakaoOAuthDTO) {
        final String provider = "kakao";

        String accessToken = oauthService.getAccessToken(kakaoOAuthDTO);
        try {
            SnsUserDTO snsUserDTO = oauthService.getKakaoUser(accessToken);
            UserEntity user = userRepository.findByProviderAndSnsId(provider, snsUserDTO.getId()
                    .toString());

            if (user != null) {
                return user;
            }
            user = snsUserDTO.toEntity(provider);
            return userRepository.save(user);
        } catch (Exception e) {
            return null;
        }
    }
}

File: backend-aspnet/Todo/Service/UserService.cs
using Microsoft.EntityFrameworkCore;
using Todo.Entity;
using Todo.Model;

namespace Todo.Service
{
    public class UserService
    {
        private readonly ILogger<UserService> _logger;
        private readonly TodoDbContext _todoDbContext;
        private readonly KakaoOAuthService _kakaoOAuthService;

        public UserService(ILogger<UserService> logger, TodoDbContext todoDbContext, KakaoOAuthService kakaoOAuthService)
        {
            _logger = logger;
            _todoDbContext = todoDbContext;
            _kakaoOAuthService = kakaoOAuthService;
        }

        public async Task<User> Create(User user)
        {
            const string provider = "local";

            if (await _todoDbContext.User.AnyAsync(e => e.Email == user.Email && e.Provider == provider) == true)
            {
                throw new Exception($"Email already exists {user.Email}");
            }

            _todoDbContext.Add(user);
            await _todoDbContext.SaveChangesAsync();

            return user;
        }

        public async Task<User?> SignInLocal(string email, string password)
        {
            const string provider = "local";

            var user = await _todoDbContext.User.FirstOrDefaultAsync(e => e.Email == email && e.Provider == provider);
            if (user != null && user.Password!.Equals(password))
                return user;
            return null;
        }

        public async Task<User?> GetByKakao(KakaoOAuthRequest request)
        {
            const string provider = "kakao";

            try
            {
                string accessToken = await _kakaoOAuthService.GetAccessToken(request);
                var kakaoUser = await _kakaoOAuthService.GetKakakoUser(accessToken);
                var user = await _todoDbContext.User.FirstOrDefaultAsync(e => e.Provider == provider && e.SnsId == kakaoUser!.Id.ToString());
                if (user == null)
                {
                    user = new User
                    {
                        Email = kakaoUser!.Email,
                        Nick = kakaoUser!.Nickname,
                        Provider = provider,
                        SnsId = kakaoUser!.Id.ToString(),
                    };
                    _todoDbContext.Add(user);
                    await _todoDbContext.SaveChangesAsync();
                }
                return user;
            }
            catch (Exception e)
            {
                _logger.LogError($"GetByKakao. exception: {e.Message}");
            }
            return null;
        }
    }
}